### ISSUE TITLE
How to run a simple python function, with inputs & outputs, from blueprints?

### ISSUE THREAD

**Issue by TylerDuncan** [20tab/UnrealEnginePython#889]
I searched the forum for this issue specifically but didn't see any answers that seemed to fit. 

If I wanted to write a python function that took an integer as input, squared it, then spit out the new integer as an output, how do I call this from a blueprint node? The current execute python method is just a string input for path with no other inputs or outputs.

Cheers,
Tyler

**Comment by tempaccountforissue** [comment]
I was able to use the 'Execute Python String' in conjuction with the 'Format Text' node to get my python script running inside of my blueprint. Just click add pin on the 'Format Text' node and you can have as many variables as you want. However, I still have not been able to figure out how to get any output from this method. One potential solution I haven't tested yet is to create a CSV file from the python script containing output data, but I'm not sure if this would work inside of a packaged project. If you are able to test this method, could you let me know if it works?

**Comment by TylerDuncan** [comment]
Ah! So you’re passing in the arguments by calling the method with the arguments in the same string? That makes sense now that I think about it. 

I did think about using text files or CSV for the output data, but it feels like there has to be a better way to pass data back.

I suppose I could place the code on an actor, and run the execute python string for the input. Then create a blueprint function that takes inputs, and call that function from the python script to pass the data back? I may be creating issues with this process but it’s worth a shot.

**Comment by tempaccountforissue** [comment]
Creating new files for output definitely isn’t the cleanest solution. I did find this example for creating variables in a blueprint from python, but I'm not sure how to use the variables in a blueprint after creating them at runtime.

https://github.com/20tab/UnrealEnginePython/blob/master/examples/blueprint_variables.py

**Comment by tempaccountforissue** [comment]
@TylerDuncan Hey, just wanted to follow up and see if you were able to do what you were trying to do. I was able to get a variable added to my blueprint, but that's no good for a runtime solution. Do you know if there's any way to set the value of a variable from python?

**Comment by TylerDuncan** [comment]
I took another stab at it tonight and got farther! I've successfully been able to call a function in the python script, pass an argument to the function (number as a string), run the function that multiplies the number by itself, then print to console,  and print screen, the output of multiplication. 

My next tests will be:
- See if I can import a python library and then print to string some output from using the library. 
- See if I can create a custom blueprint node that lets me pass back more than just a string or bool. These "execute actor method" functions are written in the plugin so maybe I can replicate them and add more output types.
- If I can't pass back anything but a string, then I'll test passing back multiple arguments as a combined string and separating them in blueprint code, then try using them.

To answer your question: I believe I have seen a way to set the value of a variable from blueprint code. It's either in the examples or the docs on github. Depending on if you're creating the actor from code then referencing it, or getting the actor in python via some find by class method.  I haven't successfully done this yet, because I'm still not able to get a reference to the actor properly in python. I just tried for half an hour but didn't get anywhere. I see there are "set python attr int" and other nodes for python character or component. I'm trying from a PyActor but those nodes don't exist there.

**Comment by tempaccountforissue** [comment]
@TylerDuncan It sounds like we’re both struggling with getting a reference. I think either “unreal_engine.editor_select_actor(an_actor)” or “ue.editor_get_selected_actors()[0]” should be able to reference an actor. For my workflow, I’m not using an actor. I have my python script running from a Widget blueprint, and I’m trying to reference and modify a variables default value in a different blueprint. You wouldn't happen to remember which example shows how to modify a variable, would you?

**Comment by TylerDuncan** [comment]
@tempaccountforissue Agreed. If I can figure out how to pass a proper reference, I think that would solve a lot of overall issues. 

I believe this is the document I'm referring to. It requires you to use it as a uobject which I'm hoping isn't editor only. Do you have any idea?

https://github.com/20tab/UnrealEnginePython/blob/master/docs/uobject_API.md

**Comment by tempaccountforissue** [comment]
@TylerDuncan I'm honestly really new to Unreal and really new to Python, so I have no idea if it would be editor only or not. I did notice this note in the doc: "NOTE: currently structs are not supported". Unfortunately, I'm trying to work with a struct array containing struct vectors so I don't think this will be able to work for me.

**Comment by TylerDuncan** [comment]
@tempaccountforissue I think you can break the struct up and pass it back and forth as a string if absolutely necessary, it would just be more tedious to have to convert it back and forth. I also don't know if that doc is old and maybe they are supported now?

**Comment by tempaccountforissue** [comment]
@TylerDuncan Yeah, I had a similar idea. My primary data is a vector which is just a stuct of floats, so if I'm able to get the floats in there then something should be possible from the blueprint side of things. I'm not really sure how to work with it as a uobject, though? I'm able to reference my blueprint with "ue.load_object(Blueprint, '/Game/WidgetBP')", but I'm not really sure where to go from there.

**Comment by tempaccountforissue** [comment]
@TylerDuncan Hey, long time no see. I was using a temporary solution to get output from my python script in the editor, but now that I'm trying to packaging my project it is no longer working. I was wonderig if you ever figured out a way to get python output inside of a blueprint?

**Comment by TylerDuncan** [comment]
@tempaccountforissue Unfortunately, I never figured out how to get this to properly work. I think I lack the understanding of c++ to do so. Instead I’m attempting to pay someone on fiverr to just create me a c++ library that will hopefully do what I wanted to do in Python, natively in c++. If you ever figure it out, let me know!

**Comment by tempaccountforissue** [comment]
@TylerDuncan The way I solved it in the editor was by writing to the default values of the variables in a seperate blueprint that isn't loaded into memory (this is important because saving a blueprint that is loaded into memory crashes the engine). Then I used the 'Get Class Defaults' node to bring the variables from the other blueprint to the one I needed them in.

Python code I used for this:
`bp = ue.load_object(Blueprint, '/Game/FooActor1679359081')
bp.NewVariables[0].DefaultValue = value
ue.compile_blueprint(bp)
bp.save_package()`

When I wrote my previous comment, I thought my script wasn't working because it wasn't able to access the blueprint after it's packaged into a pak file. However, I just tried tried to do a simple print statement to see if it was able to access my python modules, and it doesn't seem to be able to locate them.

Print statement I tested:
`import unreal_engine as ue
print(ue.get_asset('/Game/FooActor1679359081'))`

The above code segment returns this in the editor:
`LogPython: <unreal_engine.UObject '/Game/FooActor1679359081' (0x00000500107ABE00) UClass 'Package' (refcnt: 2)>`

However, it returns this error in a packaged project:
`LogPython: Error: AttributeError: module 'unreal_engine' has no attribute 'get_asset'`

I believe the reason for this has something to do with the ue_site.py file mentioned in the readme. I have tried adding a ue_site.py file in the Content/Scripts/ directory, but I keep getting this error:
`LogPython: ue_site Python module not found`

If you could help me with setting up the ue_site.py file I think we'd be able to solve each others problems (and save you some money as well).
